lib/rules.js

```diff
@@ -184,6 +188,179 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
+// e.g. nthRoot(4, 2) * nthRoot(x^2, 2) -> 2 * x
+export const EVALUATE_DISTRIBUTED_NTH_ROOT = defineRule(
+    (node) => {
+        let isDistributed = query.isMul(node)
+        let canEvaluate = false
+        if (isDistributed) {
+            canEvaluate = node.args.every(arg => canApplyRule(CANCEL_EXPONENT, arg) || canApplyRule(NTH_ROOT_VALUE, arg))
+        }
+        return (isDistributed && canEvaluate) ? {node} : null
+    },
+
+    (node) => {
+        const result = build.applyNode(
+            'mul',
+            node.args.map(nthRoot => {
+                const [radicand, index] = nthRoot.args
+                if (query.isNumber(radicand)) {
+                    return applyRule(NTH_ROOT_VALUE,
+                                     build.applyNode(
+                                         'nthRoot',
+                                         [radicand, index]
+                                     ))
+                } else {
+                    return applyRule(CANCEL_EXPONENT,
+                                     build.applyNode(
+                                         'nthRoot',
+                                         [radicand, index]
+                                     ))
+                }
+            })
+        )
+        return result
+    }
+)
+
+// e.g. 12 -> 2 * 2 * 3
+export const FACTOR_INTO_PRIME = defineRule(
+    (node) => {
+        return query.isNumber(node) ? {node} : null
+    },
+
+    (node) => {
+        const factors = primeFactorization(query.getValue(node))
+        return build.applyNode('mul', factors.map(build.numberNode))
+    }
+)
+
+// e.g. nthRoot(2 * 2 * 2, 2) -> nthRoot((2 * 2) * 2, 2)
+export const GROUP_TERMS_BY_ROOT = defineRule(
+    (node) => {
+        const canGroupTerms = canApplyRule(DISTRIBUTE_NTH_ROOT, node)
+        return canGroupTerms ? {node} : null
+    },
+
+    (node) => {
+        const radicand = node.args[0]
+        const index = node.args[1]
+
+        // dictionary storing the number of times a constant appears
+        // e.g. 2 * 2 * 2 => {'2': 3} , 2 appears 3 times
+        const count = {}
+        radicand.args.forEach(arg => {
+            const key = JSON.stringify(arg)
+            count[key] ? count[key]++ : count[key] = 1
+        })
+
+        const flatten = arr => arr.reduce(
+            (acc, val) => acc.concat(
+                Array.isArray(val) ? flatten(val) : val
+            ),
+            []
+        )
+
+        const result = build.applyNode(
+            'nthRoot',
+            [build.applyNode(
+                'mul',
+                flatten(Object.keys(count).map(key => {
+                    let leftover = count[key]
+                    const term = JSON.parse(key)
+                    const times = query.getValue(index)
+
+                    const args = []
+
+                    while (leftover - times > 0) {
+                        leftover -= times
+                        args.push(Array(times).fill(term))
```

Comment:
Instead of adding an array here and then removing it later... we should create the `mul` node right here, e.g.
```
args.push(build.applyNode('mul', Array(times).fill(term)))
```